include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -0,0 +1,1043 @@
+     * @p ADNumberType template parameter is indeed a supported
+     * auto-differentiable number.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template<typename ADNumberType, typename = void>
+    struct ADNumberTraits;
+
+
+    /**
+     * This namespace defines the classes that help provide a unified interface
+     * to all auto-differentiation numbers.
+     */
+    namespace internal
+    {
+
+      // The following three classes, namely ADNumberInfoFromEnum, Marking, and
+      // ExtractData, are those that need to be implemented for each new
+      // auto-differentiable number type. This information is then used by
+      // NumberTraits and ADNumberTraits to provide a uniform interface, as used
+      // by our drivers, to the underlying number types.
+
+
+      /**
+       * A struct that defines some fundamental information about a
+       * auto-differentiable number based on the @p ScalarType and the
+       * AD-enumeration selected by @p ADNumberTypeCode. This information is
+       * used in other convenience classes and templated functions to
+       * automatically determine information about the auto-differentiable
+       * number that has been selected to wrap the @p ScalarType.
+       *
+       * The specializations of this class have to implement the following
+       * member data and type definitions:
+       * @code
+       *
+       *   // State whether the auto-differentiable number uses taping or not.
+       *   static const bool             is_taped;
+       *   // The real-type for the auto-differentiable number
+       *   typedef <ADNumberType>        real_type;
+       *   // The type of number returned when taking the first derivative of the @p real_type.
+       *   typedef <Scalar/ADNumberType> derivative_type;
+       *   // The number of derivative levels computable from the @p real_type.
+       *   static const unsigned int     n_supported_derivative_levels;
+       *
+       * @endcode
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template<typename ScalarType, enum NumberTypes ADNumberTypeCode, typename = void>
+      struct ADNumberInfoFromEnum;
+
+
+      /**
+       * A struct to assist with the marking of AD numbers that represent
+       * independent and dependent variables.
+       *
+       * The specializations of this class have to implement the following
+       * member functions:
+       * @code
+       *
+       *  // Initialize the state of an independent variable.
+       *  static void
+       *  independent_variable(ADNumberType       &out,
+       *                       const ScalarType   &in,
+       *                       const unsigned int &index,
+       *                       const unsigned int &n_independent_variables)
+       *
+       *  // Initialize the state of a dependent variable.
+       *  static void
+       *  dependent_variable(ADNumberType       &out,
+       *                     const ADNumberType &func);
```

Comment:
same here

Reply:
Arguments to the aforementioned static member functions have been swapped both here in the documentation and as necessary in the `Marking `classes.